helix-term/src/commands/typed.rs

```diff
@@ -108,30 +114,115 @@ fn open(cx: &mut compositor::Context, args: &[Cow<str>], event: PromptEvent) ->
+        let path_str = path.to_str().context("invalid unicode")?;
+        GlobBuilder::new(path_str)
+            .literal_separator(true)
+            .empty_alternates(true)
+            .build()
+            .context("invalid glob")
+    }
+
     for arg in args {
         let (path, pos) = args::parse_file(arg);
-        let path = helix_stdx::path::expand_tilde(path);
-        // If the path is a directory, open a file picker on that directory and update the status
-        // message
-        if let Ok(true) = std::fs::canonicalize(&path).map(|p| p.is_dir()) {
-            let callback = async move {
-                let call: job::Callback = job::Callback::EditorCompositor(Box::new(
-                    move |editor: &mut Editor, compositor: &mut Compositor| {
-                        let picker = ui::file_picker(path.into_owned(), &editor.config());
-                        compositor.push(Box::new(overlaid(picker)));
-                    },
-                ));
-                Ok(call)
+        let path = helix_stdx::path::canonicalize(path);
+        if let Ok(metadata) = path.metadata() {
+            // Path exists
+            // Shortcut for opening a path without globbing
+            if metadata.is_dir() {
+                // If the path is a directory, open a file picker on that directory
+                let callback = async move {
+                    let call: job::Callback = job::Callback::EditorCompositor(Box::new(
+                        move |editor: &mut Editor, compositor: &mut Compositor| {
+                            let picker = ui::file_picker(path, &editor.config());
+                            compositor.push(Box::new(overlaid(picker)));
+                        },
+                    ));
+                    Ok(call)
+                };
+                cx.jobs.callback(callback);
+            } else {
+                open_file(cx, &path, pos)?;
+            }
+            continue;
+        }
+
+        let mut glob_set_builder = GlobSet::builder();
+        glob_set_builder.add(glob(&path)?);
+
+        let mut root = None;
+        let mut comps = path.components();
+
+        // Iterate over all parents
+        while comps.next_back().is_some() {
+            let parent = comps.as_path();
+            if parent.as_os_str().is_empty() {
+                // No parents left
+                break;
+            }
+
+            // Add each parent as a glob for filtering in the recursive walker
+            glob_set_builder.add(glob(parent)?);
+
+            if root.is_none() && parent.exists() {
+                // Found the first parent that exists
+                root = Some(parent.to_path_buf());
+            }
+        }
+
+        let glob_set = glob_set_builder.build().context("invalid glob")?;
+        let root = root.unwrap_or_else(current_working_dir);
+
+        let mut walk = cx
+            .editor
+            .config()
+            .file_picker
+            .walk_builder(root)
```

Comment:
Don't reuse the walk idler config of the file picker here. We do not want to be ignoring files here, in fact make sure to turn all default options for ignore entirely. If you write a glib pattern that matches a hidden file then that hidden file should be opened